### PR TITLE
Implement subgroup reduction for SYCL RangePolicy parallel_reduce

### PR DESCRIPTION
**Pull request by masterleinad**
Using subgroups for the reductions should improve the performance for `parallel_reduce` using `SYCL`.

### CODE REVIEW

**Comment by masterleinad** [comment]
With [`154eff5` (#3940)](https://github.com/kokkos/kokkos/pull/3940/commits/154eff5195742e5da9a58b97ace95ac5309919c8) I'm seeing
```
n = 128
AXPBY KK: 4.422993e-02 s 6.468522e-02 GB/s
AXPBY SYCL: 2.562794e-02 s 1.116369e-01 GB/s
DOT KK: 6.557975e-02 s 2.908442e-02 GB/s
DOT SYCL: 8.905854e-02 s 2.141680e-02 GB/s
n = 1024
AXPBY KK: 4.248038e-02 s 5.387943e-01 GB/s
AXPBY SYCL: 2.707055e-02 s 8.455012e-01 GB/s
DOT KK: 9.947761e-02 s 1.533892e-01 GB/s
DOT SYCL: 1.804862e-01 s 8.454271e-02 GB/s
n = 8192
AXPBY KK: 5.261190e-02 s 3.480305e+00 GB/s
AXPBY SYCL: 2.776235e-02 s 6.595460e+00 GB/s
DOT KK: 1.428609e-01 s 8.544697e-01 GB/s
DOT SYCL: 2.057009e-01 s 5.934358e-01 GB/s
n = 65536
AXPBY KK: 1.451227e-01 s 1.009383e+01 GB/s
AXPBY SYCL: 2.890844e-02 s 5.067184e+01 GB/s
DOT KK: 2.128002e-01 s 4.589105e+00 GB/s
DOT SYCL: 2.522376e-01 s 3.871598e+00 GB/s
n = 524288
AXPBY KK: 1.962914e-01 s 5.970078e+01 GB/s
AXPBY SYCL: 5.099575e-02 s 2.297986e+02 GB/s
DOT KK: 2.496850e-01 s 3.128943e+01 GB/s
DOT SYCL: 4.468903e-01 s 1.748192e+01 GB/s
n = 4194304
AXPBY KK: 2.972564e-01 s 3.153843e+02 GB/s
AXPBY SYCL: 1.551411e-01 s 6.042888e+02 GB/s
DOT KK: 3.195238e-01 s 1.956036e+02 GB/s
DOT SYCL: 5.425608e-01 s 1.151945e+02 GB/s
n = 33554432
AXPBY KK: 1.244910e+00 s 6.024532e+02 GB/s
AXPBY SYCL: 1.025914e+00 s 7.310552e+02 GB/s
DOT KK: 9.931421e-01 s 5.034526e+02 GB/s
DOT SYCL: 1.252469e+00 s 3.992115e+02 GB/s
n = 268435456
AXPBY KK: 8.666444e+00 s 6.923255e+02 GB/s
AXPBY SYCL: 8.056914e+00 s 7.447020e+02 GB/s
DOT KK: 5.673008e+00 s 7.050933e+02 GB/s
DOT SYCL: 6.063481e+00 s 6.596871e+02 GB/s
```
on a V100 so we are faster than the native implementation (at least for this simple test case).

**Comment by masterleinad** [comment]
I am happy with the current status and am looking for reviews.

**Review by crtrott — CHANGES_REQUESTED** [review]
Besides the failing test, would it be practical to lift the reduction code for a group and subgroup out of here into its own file and call it from these places. It looks like its the exact same code for the three usecases (minus the copy to global for the team internal reduction). Otherwise this looks pretty clean, and I didn't see a mistake which would explain the failing tests.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
@crtrott I fixed the failing tests and factored common code out.

**Review by nliber — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 1]
```diff
@@ -52,8 +52,84 @@
...
+namespace SYCLReduction {
+template <class ValueJoin, class ValueOps, typename WorkTag, typename ValueType,
+          typename ReducerType, typename FunctorType, int dim>
+void workgroup_reduction(sycl::nd_item<dim>& item,
+                         sycl::local_ptr<ValueType> local_mem,
+                         ValueType* results_ptr, const unsigned int value_count,
+                         const ReducerType& selected_reducer,
+                         const FunctorType& functor, bool final) {
+  const auto local_id = item.get_local_linear_id();
+  // FIXME_SYCL should be item.get_group().get_local_linear_range();
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:67]: What prevents you to do so?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 2]
```diff
@@ -52,8 +52,84 @@
...
+  for (unsigned int stride = local_range / 2; stride > 0; stride >>= 1) {
+    auto* tmp = sg.shuffle_down(result, stride);
+    if (id_in_sg + stride < local_range)
+      ValueJoin::join(selected_reducer, result, tmp);
+  }
+  item.barrier(sycl::access::fence_space::local_space);
+
+  // Copy the subgroup results into the first positions of the
+  // reduction array.
+  if (sg.get_local_id()[0] == 0 && sg.get_local_id()[0] < wgroup_size)
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:90]: Why are you not using `id_in_sg` here?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp [hunk 3]
```diff
@@ -52,8 +52,84 @@
...
+    ValueOps::copy(functor, &local_mem[sg.get_group_id()[0] * value_count],
+                   result);
+  item.barrier(sycl::access::fence_space::local_space);
+
+  // Do the final reduction only using the first subgroup.
+  if (sg.get_group_id()[0] == 0) {
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:96]: Same question

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:67]: It's not implemented.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp (hunk 3 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp:96]: This is not `sg.get_local_id()[0]`.

File: core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp (hunk 2 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Parallel_Reduce.hpp]: Replaced.

**Comment by masterleinad** [comment]
I'm going to assume that we will merge this more or less as is so I can base other improvements on top of this.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**